src/compas_ags/ags/constraints.py

```diff
@@ -0,0 +1,372 @@
+
+    Parameters
+    -----------
+    form: :class:`FormDiagram`
+        The Form Diagram to constraint
+    vertex: int
+        Key of the vertex to fix.
+
+    """
+
+    def __init__(self, form, vertex):
+        super().__init__(form)
+        self.vertex = vertex
+        self.vertex_index = form.vertex_index()
+        self.P = None
+        self.set_initial_position()
+
+    def set_initial_position(self):
+        self.P = self.form.vertex_attribute(self.vertex, 'y')
+
+    def compute_constraint(self):
+        constraint_jac_row = np.zeros((1, self.number_of_cols))
+        idx = self.vertex_index[self.vertex] + self.vcount
+        constraint_jac_row[0, idx] = 1
+        r = self.form.vertex_attribute(self.vertex, 'y') - self.P
+        return constraint_jac_row, r
+
+    def update_constraint_goal(self):
+        self.set_initial_position()
+
+    def get_lines(self):
+        constraint_lines = []
+        s = self.form.vertex_coordinates(self.vertex, 'xy')
+        e = self.form.vertex_coordinates(self.vertex, 'xy')
+        s[0] += 1
+        e[0] -= 1
+        constraint_lines.append({
+            'start': s,
+            'end': e,
+            'width': self._width,
+            'color': self._color,
+            'style': self._style,
+        })
+        return constraint_lines
+
+
+class AngleFix(AbstractConstraint):
+    """Constraint that keeps vertex fixed along an inclined line.
+
+    Parameters
+    -----------
+    form: :class:`FormDiagram`
+        The Form Diagram to constraint
+    vertex: int
+        Key of the vertex to fix.
+    angle: float
+        Angle (clockwise) to fix the vertex to.
+
+    """
+
+    def __init__(self, form, vertex, angle):
+        super().__init__(form)
+        self.vertex = vertex
+        self.angle = angle
+        self.vertex_index = form.vertex_index()
+        self.x = None
+        self.y = None
+        self.set_initial_position()
+
+    def set_initial_position(self):
+        self.x = self.form.vertex_attribute(self.vertex, 'x')
+        self.y = self.form.vertex_attribute(self.vertex, 'y')
+
+    def compute_constraint(self):
+        constraint_jac_row = np.zeros((1, self.number_of_cols))
+
+        idx = self.vertex_index[self.vertex]
+        constraint_jac_row[0, idx] = 1 * math.sin(math.radians(self.angle))
+        r = (self.form.vertex_attribute(self.vertex, 'x') - self.x) * math.sin(math.radians(self.angle))
+
+        idx = self.vertex_index[self.vertex] + self.vcount
+        constraint_jac_row[0, idx] = 1 * math.cos(math.radians(self.angle))
+        r = (self.form.vertex_attribute(self.vertex, 'y') - self.y) * math.cos(math.radians(self.angle))
+        return constraint_jac_row, r
+
+    def update_constraint_goal(self):
+        self.set_initial_position()
+
+    def get_lines(self):
+        constraint_lines = []
+        s = self.form.vertex_coordinates(self.vertex, 'xy')
+        e = self.form.vertex_coordinates(self.vertex, 'xy')
```

Comment:
s and e are the same?

Reply:
sort of, they have coordinates changed in different ways afterwards to generate the lines in the viewer representing the constraint.